File: regression/cs/tst16.cs
using Microsoft.Z3;
using System;

class Test
{

    public void Run()
    {        
        using (Context ctx = new Context())
        {
            Sort U = ctx.MkUninterpretedSort("U");
            FuncDecl f = ctx.MkFuncDecl("f", U, U);
            Expr a = ctx.MkConst("a", U);
            Expr b = ctx.MkConst("b", U);
            Expr c = ctx.MkConst("c", U);

            Solver s = ctx.MkSolver();
            s.Assert(ctx.MkEq(f[f[a]], b),
                     ctx.MkNot(ctx.MkEq(f[b], c)),
                     ctx.MkEq(f[c], c));
            Console.WriteLine(s.Check());
            Model m = s.Model;
            foreach (FuncDecl d in m.Decls)
                if (d.DomainSize == 0)
                    Console.WriteLine(d.Name + " -> " + m.ConstInterp(d));
                else 
                    Console.WriteLine(d.Name + " -> " + m.FuncInterp(d));
            
            Console.WriteLine(m.NumSorts);
            Console.WriteLine(m.Sorts[0]);

            foreach(Sort srt in m.Sorts)
                Console.WriteLine(srt);
            
            foreach (Expr v in m.SortUniverse(U))
                Console.WriteLine(v);
        }
    }
}

File: old-regressions/python/tst16.py
from z3 import *

U = DeclareSort('U')
f = Function('f', U, U)
a, b, c = Consts('a b c', U)
s = Solver()
s.add(f(f(a)) == b, f(b) != c, f(c) == c)
print s.check()
m = s.model()
for d in m:
    print "%s -> %s" % (d.name(), m[d])
print m.num_sorts()
print m.get_sort(0)
print m.sorts()
print m[U]
print "universe for U:"
for v in m[U]:
    print v
